openmls/src/group/mls_group/mod.rs

```diff
@@ -72,50 +72,107 @@ implement_persistence!(
-        version: impl Into<Option<ProtocolVersion>>,
-    ) -> Result<Self, MlsGroupError> {
-        debug!("Created group {:x?}", id);
-        trace!(" >>> with {:?}, {:?}", ciphersuite_name, config);
-        let group_id = GroupId { value: id.into() };
-        let ciphersuite = Config::ciphersuite(ciphersuite_name)?;
-        let tree = RatchetTree::new(backend, key_package_bundle);
-        // TODO #483: Implement extensions
-        let extensions: Vec<Extension> = Vec::new();
+/// Builder for [`MlsGroup`].
+pub struct MlsGroupBuilder {
+    key_package_bundle: KeyPackageBundle,
+    group_id: Option<GroupId>,
+    config: Option<MlsGroupConfig>,
+    psk: Option<PskSecret>,
+    version: Option<ProtocolVersion>,
+    required_capabilities: Option<RequiredCapabilitiesExtension>,
+}
+
+impl MlsGroupBuilder {
+    /// Create a new [`MlsGroupBuilder`].
+    pub fn new(key_package_bundle: KeyPackageBundle) -> Self {
+        Self {
+            key_package_bundle,
+            group_id: None,
+            config: None,
+            psk: None,
+            version: None,
+            required_capabilities: None,
+        }
+    }
+    /// Set the [`GroupId`] of the [`MlsGroup`].
+    pub fn with_group_id(mut self, group_id: GroupId) -> Self {
+        self.group_id = Some(group_id);
+        self
+    }
+    /// Set the [`GroupId`] of the [`MlsGroup`].
+    pub fn with_group_id_slice(mut self, group_id: &[u8]) -> Self {
+        self.group_id = Some(GroupId::from_slice(group_id));
+        self
+    }
+    /// Set the [`MlsGroupConfig`] of the [`MlsGroup`].
+    pub fn with_config(mut self, config: MlsGroupConfig) -> Self {
+        self.config = Some(config);
+        self
+    }
+    /// Set the [`PskSecret`] of the [`MlsGroup`].
+    pub fn with_psk(mut self, psk: PskSecret) -> Self {
+        self.psk = Some(psk);
+        self
+    }
+    /// Set the [`ProtocolVersion`] of the [`MlsGroup`].
+    pub fn with_version(mut self, version: ProtocolVersion) -> Self {
+        self.version = Some(version);
+        self
+    }
+    /// Set the [`RequiredCapabilitiesExtension`] of the [`MlsGroup`].
+    pub fn with_required_capabilities(
+        mut self,
+        required_capabilities: RequiredCapabilitiesExtension,
+    ) -> Self {
+        self.required_capabilities = Some(required_capabilities);
+        self
+    }
+
+    /// Build the [`MlsGroup`].
+    /// Any values that haven't been set in the builder are set to their default
+    /// values (which might be random).
+    ///
+    /// This function performs cryptographic operations and there requires an
+    /// [`OpenMlsCryptoProvider`].
+    pub fn build(self, backend: &impl OpenMlsCryptoProvider) -> Result<MlsGroup, MlsGroupError> {
+        let ciphersuite = self.key_package_bundle.key_package().ciphersuite();
+        let config = self.config.unwrap_or_default();
+        let group_id = self.group_id.unwrap_or_else(|| GroupId::random(backend));
```

Comment:
I think this is potentially dangerous. I'd prefer it if we made the group ID mandatory.

Reply:
I don't think it's dangerous but I don't have strong opinions on this. I'll make group ID mandatory.